File: src/tools/fs-tools.ts
import { readFile as readFileRaw } from "node:fs/promises";
import fg from "fast-glob";

type ReadFileOptions = {
  encoding?: BufferEncoding;
};

type GlobOptions = {
  cwd?: string;
  onlyFiles?: boolean;
  absolute?: boolean;
};

// Thin wrapper over Node's fs — used by loadEvidence for consistent error handling
export async function readFile(path: string, options: ReadFileOptions = {}): Promise<string> {
  const encoding = options.encoding ?? "utf8";
  return readFileRaw(path, encoding);
}

// Expand glob patterns to absolute file paths via fast-glob
export async function glob(patterns: string[], options: GlobOptions = {}): Promise<string[]> {
  return fg(patterns, {
    cwd: options.cwd,
    onlyFiles: options.onlyFiles ?? true,
    absolute: options.absolute ?? true,
  });
}

// Case-insensitive line filter — used for quick keyword searches in evidence files
export function grepLines(content: string, query: string): string[] {
  const lines = content.split(/\r?\n/);
  return lines.filter((line) => line.toLowerCase().includes(query.toLowerCase()));
}

File: python/grc_agent/tools/fs_tools.py
from __future__ import annotations

import glob as glob_mod
import re
from pathlib import Path

import anyio


# Thin wrapper over anyio's file I/O — used by _load_evidence for consistent error handling
async def read_file(path: str, encoding: str = "utf-8") -> str:
    """Read a file and return its contents as a string."""
    return await anyio.Path(path).read_text(encoding=encoding)


# Expand glob patterns to absolute file paths
async def glob_files(patterns: list[str], root: str | None = None) -> list[str]:
    """Expand glob patterns and return matching file paths."""
    base = root or str(Path.cwd())
    results: list[str] = []
    for pattern in patterns:
        matches = glob_mod.glob(pattern, root_dir=base, recursive=True)
        for match in matches:
            full = str(Path(base) / match)
            if full not in results:
                results.append(full)
    return results


# Case-insensitive line filter — used for quick keyword searches in evidence files
def grep_lines(content: str, query: str) -> list[str]:
    """Return lines containing *query* (case-insensitive)."""
    lower_query = query.lower()
    return [line for line in content.splitlines() if lower_query in line.lower()]
